### PR TITLE
#1: Add translation for dynamic value

### PR DESCRIPTION
**Pull request by mattnguyengit**
- [x] Solve %s for change many times
- [x] Solve %s for date time format
- Check for dynamic directive on admin modules (we will check later)

### CODE REVIEW

**Comment by mattnguyengit** [comment]
On backend, we change translation from
```
{
  "%s selected": "%s selected"
}
```
to
{
  "%s selected email": "{{ items.length }} selected email",
}

Because we have the same "%s selected" translation on other modules so  "{{ items.length }} selected email" is overrided.

**Comment by mattnguyengit** [comment]
On commit, I use Object.defineProperty to call the getSelectedItems() every time we get the length value of items.

I want to discuss it's the good approach to apply all other dynamic value with %s

**Comment by ppppbn** [comment]
> On commit, I use Object.defineProperty to call the getSelectedItems() every time we get the length value of items.
> 
> I want to discuss it's the good approach to apply all other dynamic value with %s

Please note that this solution is only for function binding directly to html. Binding variables as values does not need this approach.

File: src/linagora.esn.unifiedinbox/app/controllers.js [hunk 1]
```diff
@@ -715,6 +716,10 @@ angular.module('linagora.esn.unifiedinbox')
       };
     });
 
+    Object.defineProperty(self.selectedItems, 'items', {
+      get() { return self.getSelectedItems()},
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/app/controllers.js:720]:
```suggestion
      get() { return self.getSelectedItems() },
```

File: src/linagora.esn.unifiedinbox/views/unified-inbox/subheader.pug [hunk 2]
```diff
@@ -8,7 +8,7 @@ block left
     .inbox-selection-aware-subheader.inbox-selecting(ng-switch-when="true")
       button.btn.btn-link.btn-icon(href='', ng-click="ctrl.unselectAllItems()", title="{{ 'Cancel' | translate }}")
         i.mdi.mdi-close
-      span.title.extra-title.ellipsis {{ '%s selected' | translate:ctrl.getSelectedItems().length }}
+      span.title.extra-title.ellipsis {{ '%s selected email' | translate:ctrl.selectedItems }}
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/views/unified-inbox/subheader.pug:11]: Please keep the original translation as we discussed

File: src/linagora.esn.unifiedinbox/app/controllers.js [hunk 3]
```diff
@@ -481,6 +484,14 @@ angular.module('linagora.esn.unifiedinbox')
                 self.fixTime('toDate');
                 self.updateDateAndTime('toDate');
               }
+
+              Object.defineProperty(self.vacationDate, 'start', {
+                get() { return esnDatetimeService.format($scope.vacation.fromDate.toDate(), 'LLL') },
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/app/controllers.js:489]: Why did you take `LLL`? Is it equal to `esnDatetime:'mediumDate time'`? Why not use mediumDate constant from esnDatetimeService?

File: src/linagora.esn.unifiedinbox/app/controllers.js [hunk 4]
```diff
@@ -481,6 +484,14 @@ angular.module('linagora.esn.unifiedinbox')
...
+              })
+
+              Object.defineProperty(self.vacationDate, 'end', {
+                get() { return esnDatetimeService.format($scope.vacation.toDate.toDate(), 'LLL') },
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/app/controllers.js:493]: same here

File: src/linagora.esn.unifiedinbox/app/controllers.js [hunk 5]
```diff
@@ -481,6 +484,14 @@ angular.module('linagora.esn.unifiedinbox')
                 self.fixTime('toDate');
                 self.updateDateAndTime('toDate');
               }
+
+              Object.defineProperty(self.vacationDate, 'start', {
+                get() { return esnDatetimeService.format($scope.vacation.fromDate.toDate(), 'LLL') },
+              })
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/app/controllers.js:490]: missing `;`

File: src/linagora.esn.unifiedinbox/app/controllers.js [hunk 6]
```diff
@@ -481,6 +484,14 @@ angular.module('linagora.esn.unifiedinbox')
...
+
+              Object.defineProperty(self.vacationDate, 'end', {
+                get() { return esnDatetimeService.format($scope.vacation.toDate.toDate(), 'LLL') },
+              })
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/app/controllers.js:494]: missing `;`

File: src/linagora.esn.unifiedinbox/app/controllers.js [hunk 7]
```diff
@@ -715,6 +727,10 @@ angular.module('linagora.esn.unifiedinbox')
       };
     });
 
+    Object.defineProperty(self.selectedItems, 'items', {
+      get() { return self.getSelectedItems() },
+    })
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/app/controllers.js:732]: missing `;`

File: src/linagora.esn.unifiedinbox/app/controllers.js [hunk 8]
```diff
@@ -481,6 +484,14 @@ angular.module('linagora.esn.unifiedinbox')
                 self.fixTime('toDate');
                 self.updateDateAndTime('toDate');
               }
+
+              Object.defineProperty(self.vacationDate, 'start', {
+                get() { return esnDatetimeService.format($scope.vacation.fromDate.toDate(), 'LLL') },
+              })
+
+              Object.defineProperty(self.vacationDate, 'end', {
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/app/controllers.js:492]: IMO, better if you have a util method to 2 `defineProperty` actions for `vacationDate`. And explain what are they for in the method doc.

File: src/linagora.esn.unifiedinbox/views/unified-inbox/subheader.pug [hunk 9]
```diff
@@ -8,7 +8,7 @@ block left
     .inbox-selection-aware-subheader.inbox-selecting(ng-switch-when="true")
       button.btn.btn-link.btn-icon(href='', ng-click="ctrl.unselectAllItems()", title="{{ 'Cancel' | translate }}")
         i.mdi.mdi-close
-      span.title.extra-title.ellipsis {{ '%s selected' | translate:ctrl.getSelectedItems().length }}
+      span.title.extra-title.ellipsis {{ '%s selected' | translate:ctrl.selectedItems }}
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/views/unified-inbox/subheader.pug:11]: Why the old one binds the length of the array but you bind the array?

Comment (mattnguyengit) [reply src/linagora.esn.unifiedinbox/views/unified-inbox/subheader.pug:11]: In the translation, we will translate {{ selectedItems.length }}. I checked

File: src/linagora.esn.unifiedinbox/app/controllers.js (hunk 3 above)
Comment (mattnguyengit) [reply src/linagora.esn.unifiedinbox/app/controllers.js]: I use $filter for it. 'mediumDate + time' --> 'longDate'

File: src/linagora.esn.unifiedinbox/app/controllers.js (hunk 8 above)
Comment (mattnguyengit) [reply src/linagora.esn.unifiedinbox/app/controllers.js]: Done I use a factory for this. It's better

File: src/linagora.esn.unifiedinbox/app/providers.js [hunk 10]
```diff
@@ -327,4 +327,12 @@ angular.module('linagora.esn.unifiedinbox')
 
       return item;
     };
+  })
+
+  .factory('getDynamicTranslatedValue', function() {
```
Comment (tuanlc) [inline src/linagora.esn.unifiedinbox/app/providers.js:332]: Add doc for this please

`getDynamicTranslatedValue` is not a correct name I think

**Pull request merged**